category_theory/yoneda.lean

```diff
@@ -0,0 +1,94 @@
+   
+   Also the Yoneda lemma, `yoneda_lemma : (yoneda_pairing C) ≅ (yoneda_evaluation C)`. -/
+
+import category_theory.natural_transformation
+import category_theory.opposites
+import category_theory.types
+import category_theory.embedding
+import category_theory.natural_isomorphism
+
+namespace category_theory
+
+universes u₁ v₁ u₂
+
+variables (C : Type u₁) [𝒞 : category.{u₁ v₁} C]
+include 𝒞
+
+def yoneda : C ⥤ ((Cᵒᵖ) ⥤ (Type v₁)) := 
+{ obj := λ X, { obj := λ Y : C, Y ⟶ X,
+                map' := λ Y Y' f g, f ≫ g,
+                map_comp' := begin intros X_1 Y Z f g, ext1, dsimp at *, erw [category.assoc] end,
+                map_id' := begin intros X_1, ext1, dsimp at *, erw [category.id_comp] end },
+  map' := λ X X' f, { app := λ Y g, g ≫ f } }
+
+namespace yoneda
+@[simp] lemma obj_obj (X Y : C) : ((yoneda C) X) Y = (Y ⟶ X) := rfl
+@[simp] lemma obj_map (X : C) {Y Y' : C} (f : Y ⟶ Y') : ((yoneda C) X).map f = λ g, f ≫ g := rfl
+@[simp] lemma map_app {X X' : C} (f : X ⟶ X') (Y : C) : ((yoneda C).map f) Y = λ g, g ≫ f := rfl
+
+lemma aux_1 {X Y : Cᵒᵖ} (f : X ⟶ Y) : ((yoneda C) X).map f (𝟙 X) = ((yoneda C).map f) Y (𝟙 Y) := by obviously
+@[simp] lemma aux_2 {X Y : C} (α : (yoneda C) X ⟶ (yoneda C) Y) 
+  {Z Z' : C} (f : Z ⟶ Z') (h : Z' ⟶ X) : f ≫ α Z' h = α Z (f ≫ h) :=
+begin erw [functor_to_types.naturality], refl end
+
+instance full : full (yoneda C) := 
+{ preimage := λ X Y f, (f X) (𝟙 X) }.
+
+instance faithful : faithful (yoneda C) := 
+begin
+  fsplit, 
+  intros X Y f g p, 
+  injection p with h,
+  convert (congr_fun (congr_fun h X) (𝟙 X)) ; simp
+end
+
+/-- Extensionality via Yoneda. The typical usage would be
+```
+-- Goal is `X ≅ Y`
+apply yoneda.ext,
+-- Goals are now functions `(Z ⟶ X) → (Z ⟶ Y)`, `(Z ⟶ Y) → (Z ⟶ X)`, and the fact that these
+functions are inverses and natural in `Z`.
+```
+-/
+def ext (X Y : C)
+  (p : Π {Z : C}, (Z ⟶ X) → (Z ⟶ Y)) (q : Π {Z : C}, (Z ⟶ Y) → (Z ⟶ X))
+  (h₁ : Π {Z : C} (f : Z ⟶ X), q (p f) = f) (h₂ : Π {Z : C} (f : Z ⟶ Y), p (q f) = f) 
+  (n : Π {Z Z' : C} (f : Z' ⟶ Z) (g : Z ⟶ X), p (f ≫ g) = f ≫ p g) : X ≅ Y := 
+@preimage_iso _ _ _ _ (yoneda C) _ _ _ _ 
+  (nat_iso.of_components (λ Z, { hom := p, inv := q, }) (by tidy))
+
+-- We need to help typeclass inference with some awkward universe levels here.
+instance prod_category_instance_1 : category (((Cᵒᵖ) ⥤ Type v₁) × (Cᵒᵖ)) := category_theory.prod.{(max u₁ (v₁+1)) (max u₁ v₁) u₁ v₁} (Cᵒᵖ ⥤ Type v₁) (Cᵒᵖ)
+instance prod_category_instance_2 : category ((Cᵒᵖ) × ((Cᵒᵖ) ⥤ Type v₁)) := category_theory.prod.{u₁ v₁ (max u₁ (v₁+1)) (max u₁ v₁)} (Cᵒᵖ) (Cᵒᵖ ⥤ Type v₁) 
+end yoneda
+
+def yoneda_evaluation : (((Cᵒᵖ) ⥤ (Type v₁)) × (Cᵒᵖ)) ⥤ (Type (max u₁ v₁)) 
+  := (evaluation (Cᵒᵖ) (Type v₁)) ⋙ ulift_functor.{v₁ u₁}
+
+@[simp] lemma yoneda_evaluation_map_down (P Q : (Cᵒᵖ ⥤ Type v₁) ×  (Cᵒᵖ)) (α : P ⟶ Q) (x : (yoneda_evaluation C) P)
+ : ((yoneda_evaluation C).map α x).down = (α.1) (Q.2) ((P.1).map (α.2) (x.down)) := rfl
```

Comment:
Also `:` on the end of the previous line.

Reply:
Fixed. (Sorry, I did learn this rule at Orsay, but I forgot to check old files.)